dotCMS/src/main/java/com/dotcms/analytics/AccessTokenRenewRunnable.java

```diff
@@ -0,0 +1,109 @@
+package com.dotcms.analytics;
+
+import com.dotcms.analytics.app.AnalyticsApp;
+import com.dotcms.analytics.cache.AnalyticsCache;
+import com.dotcms.analytics.helper.AnalyticsHelper;
+import com.dotcms.analytics.model.AccessToken;
+import com.dotcms.exception.AnalyticsException;
+import com.dotcms.exception.UnrecoverableAnalyticsException;
+import com.dotmarketing.business.APILocator;
+import com.dotmarketing.business.CacheLocator;
+import com.dotmarketing.quartz.job.AccessTokenRenewJob;
+import com.dotmarketing.util.Logger;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Access Token renew thread.
+ * Meant to perform several attempts to renew {@link AccessToken} when is required to.
+ *
+ * @author vico
+ */
+public class AccessTokenRenewRunnable implements Runnable {
+
+    private final AccessTokenRenewJob callingJob;
+    private final List<AnalyticsApp> analyticsApps;
+    private final AnalyticsAPI analyticsAPI;
+    private final AnalyticsCache analyticsCache;
+
+    public AccessTokenRenewRunnable(final AccessTokenRenewJob callingJob, final List<AnalyticsApp> analyticsApps) {
+        this.callingJob = callingJob;
+        this.analyticsApps = analyticsApps;
+        analyticsAPI = APILocator.getAnalyticsAPI();
+        analyticsCache = CacheLocator.getAnalyticsCache();
+    }
+
+    /**
+     * Run access token renew thread's main logic.
+     */
+    @Override
+    public void run() {
+        final String clientIds = analyticsApps
+            .stream()
+            .map(app -> app.getAnalyticsProperties().clientId())
+            .collect(Collectors.joining(","));
+        if (callingJob.isRenewRunning()) {
+            Logger.debug(
+                this,
+                String.format(
+                    "Renewing thread for clientIds %s is already running, skipping this execution",
+                    clientIds));
+            return;
+        }
+
+        callingJob.setRenewRunning(true);
```

Comment:
it seem it is such as a flag state, may use a try finally here?

Reply:
Done.